tools/fuzzer/expr_gen.cc

```diff
@@ -157,22 +144,52 @@ Expr ExprGenerator::maybe_parenthesized(Expr expr) {
 Expr ExprGenerator::generate() {
   WeightsArray weights;
   for (size_t i = 0; i < weights.size(); i++) {
-    weights[i] = EXPR_KIND_INFO[i].initial_weight;
+    weights[i] = cfg_.expr_kind_weights[i].initial_weight;
   }
 
   return gen_with_weights(weights);
 }
 
-BinOp DefaultGeneratorRng::gen_bin_op() {
-  std::uniform_int_distribution<int> distr((int)BinOp::EnumFirst,
-                                           (int)BinOp::EnumLast);
-  return (BinOp)distr(rng_);
+BinOp DefaultGeneratorRng::gen_bin_op(const BinOpMask& mask) {
+  assert(mask.any());
+
+  std::uniform_int_distribution<size_t> distr(0, mask.count() - 1);
+  size_t choice = distr(rng_);
+
+  size_t running_ones = 0;
+  for (size_t i = 0; i < mask.size(); i++) {
+    if (running_ones == choice && mask[i]) {
+      return (BinOp)i;
+    }
+    if (mask[i]) {
+      running_ones++;
+    }
+  }
+
+  // Fallback, should never be reached
+  assert(false);
+  return BinOp::LogicalAnd;
 }
 
-UnOp DefaultGeneratorRng::gen_un_op() {
-  std::uniform_int_distribution<int> distr((int)UnOp::EnumFirst,
-                                           (int)UnOp::EnumLast);
-  return (UnOp)distr(rng_);
+UnOp DefaultGeneratorRng::gen_un_op(const UnOpMask& mask) {
+  assert(mask.any());
+
+  std::uniform_int_distribution<size_t> distr(0, mask.count() - 1);
+  size_t choice = distr(rng_);
+
+  size_t running_ones = 0;
+  for (size_t i = 0; i < mask.size(); i++) {
+    if (running_ones == choice && mask[i]) {
+      return (UnOp)i;
+    }
+    if (mask[i]) {
+      running_ones++;
+    }
+  }
```

Comment:
Alternative way to compute this. I'm not sure which one is easier to understand (mine seems a bit more intuitive to me, but dunno), so up to you to decide.

```suggestion
  std::uniform_int_distribution<size_t> distr(1, mask.count());
  size_t choice = distr(rng_);

  size_t running_ones = 0;
  for (size_t i = 0; i < mask.size(); i++) {
    if (mask[i]) {
      running_ones++;
    }
    if (running_ones == choice) {
      return (UnOp)i;
    }
  }
```